shell/platform/android/io/flutter/embedding/engine/android/FlutterSurfaceView.java

```diff
@@ -0,0 +1,159 @@
+
+/**
+ * Paints a Flutter UI on a {@link android.view.Surface}.
+ *
+ * Consider using a {@link FlutterView}, which utilizes a {@code FlutterSurfaceView}
+ * internally.
+ */
+public class FlutterSurfaceView extends SurfaceView {
+  private static final String TAG = "FlutterSurfaceView";
+
+  private boolean isSurfaceAvailableForRendering = false;
+  private boolean isAttachedToFlutterRenderer = false;
+  @Nullable
+  private FlutterRenderer flutterRenderer;
+
+  // Connects the {@code Surface} beneath this {@code SurfaceView} with Flutter's native code.
+  // Callbacks are received by this Object and then those messages are forwarded to our
+  // FlutterRenderer, and then on to the JNI bridge over to native Flutter code.
+  private final SurfaceHolder.Callback surfaceCallback = new SurfaceHolder.Callback() {
+    @Override
+    public void surfaceCreated(SurfaceHolder holder) {
+      Log.d(TAG, "SurfaceHolder.Callback.surfaceCreated()");
+      isSurfaceAvailableForRendering = true;
+
+      if (isAttachedToFlutterRenderer) {
+        Log.d(TAG, "Already attached to renderer. Notifying of surface creation.");
+        connectSurfaceToRenderer();
+      }
+    }
+
+    @Override
+    public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
+      if (isAttachedToFlutterRenderer) {
+        changeSurfaceSize(width, height);
+      }
+    }
+
+    @Override
+    public void surfaceDestroyed(SurfaceHolder holder) {
+      Log.d(TAG, "SurfaceHolder.Callback.surfaceDestroyed()");
+      isSurfaceAvailableForRendering = false;
+
+      if (isAttachedToFlutterRenderer) {
+        disconnectSurfaceFromRenderer();
+      }
+    }
+  };
+
+  /**
+   * Constructs a {@code FlutterSurfaceView} programmatically, without any XML attributes.
+   */
+  public FlutterSurfaceView(Context context) {
+    this(context, null);
+  }
+
+  /**
+   * Constructs a {@code FlutterSurfaceView} in an XML-inflation-compliant manner.
+   */
+  public FlutterSurfaceView(Context context, AttributeSet attrs) {
+    super(context, attrs);
+    init();
+  }
+
+  private void init() {
+    // Grab a reference to our underlying Surface and register callbacks with that Surface so we
+    // can monitor changes and forward those changes on to native Flutter code.
+    getHolder().addCallback(surfaceCallback);
+  }
+
+  /**
+   * Invoked when this {@code FlutterSurfaceView} is connected to a {@link FlutterRenderer} as
+   * a {@link FlutterRenderer.RenderSurface}.
+   *
+   * If an Android {@link android.view.Surface} is available, this method will begin rendering
+   * {@link FlutterRenderer}'s Flutter UI to this {@code FlutterSurfaceView}.
+   *
+   * If no Android {@link android.view.Surface} is available yet, this {@code FlutterSurfaceView}
+   * will wait until a {@link android.view.Surface} becomes available and then begin rendering.
+   */
+  public void onAttachedToRenderer(@NonNull FlutterRenderer flutterRenderer) {
```

Comment:
Should this be public if it's only meant to be invoked as part of an internal callback? What happens if somebody tries to call this in their app directly?

Same question for the other public callbacks below and in `TextureView`.

Reply:
I'm not sure what you mean by internal callback.  This method is essentially a lifecycle method that we're introducing.  If a developer would like to utilize a `FlutterSurfaceView` for his/her own purposes then that developer must connect it to a `FlutterRenderer`.  That's done here.  Whoever owns an instance of this class is responsible for correctly invoking these lifecycle methods, just like how the Android OS has to invoke Activity lifecycle methods.

Reply:
Ah got it, I misunderstood the purpose of this. I was confused by the passive verb tense in this line:

> `Invoked when this {@code FlutterSurfaceView} is connected to a {@link FlutterRenderer} as a {@link FlutterRenderer.RenderSurface}.`

So it's not that these are callbacks that are automatically invoked when the condition is met. These are lifecycle methods that need to be manually invoked by the owner of an instance of this class during the given conditions. Can you reword the docs to explain that responsibility more clearly? Something like:

> `This lifecycle method must be called when this is connected to a {@link FlutterRenderer} as a {@link FlutterRenderer.RenderSurface}.`

I think it may make it clearer if there were an interface defining these lifecycle callbacks and when they should be manually called separately that this and `TextureView` both implemented. That way everything would be clearly grouped together so it would be easier for users of this to track what they needed to wire in where. But that's again subjective.

Reply:
I took another crack at it and also added some related comments to the class javadocs.

Reply:
WRT to the interface comment - give me a few more PRs to get there.  I think the right answer is for me to refactor the concept of a RenderSurface which I've already introduced.  I don't think RenderSurface is serving the correct purpose right now - it should probably become the interface you're talking about here.

Reply:
I think the passive verb tense here in the summary fragment is still misleading but it's a minor nit.